src/back/wgsl/writer.rs

```diff
@@ -1169,15 +1196,28 @@ impl<W: Write> Writer<W> {
                 write!(self.out, ")")?;
             }
             Expression::Load { pointer } => {
-                let is_atomic = match *func_ctx.info[pointer].ty.inner_with(&module.types) {
-                    crate::TypeInner::Pointer { base, .. } => match module.types[base].inner {
-                        crate::TypeInner::Atomic { .. } => true,
-                        _ => false,
-                    },
-                    _ => false,
+                let (mut is_pointer, is_atomic) =
+                    match *func_ctx.info[pointer].ty.inner_with(&module.types) {
+                        crate::TypeInner::Pointer { base, .. } => (
+                            true,
+                            match module.types[base].inner {
+                                crate::TypeInner::Atomic { .. } => true,
+                                _ => false,
+                            },
+                        ),
+                        _ => (false, false),
+                    };
+                is_pointer &= match func_ctx.expressions[pointer] {
+                    // Variables have a pointer return type but shouldn't be dereferenced
+                    Expression::LocalVariable(_) | Expression::GlobalVariable(_)
+                    // Access chains might have pointer types but glsl considers them as references
+                    | Expression::AccessIndex {..} | Expression::Access {..} => false,
+                    _ => true,
                 };
                 if is_atomic {
                     write!(self.out, "atomicLoad({}", ATOMIC_REFERENCE)?;
```

Comment:
do we still need `ATOMIC_REFERENCE`, given that your change fixes the pointers?

Reply:
The wgsl spec doesn't state it clearly but I think that global variables return a `ref` instead of a `ptr` so we would still need this, but it's really hard to figure out what the spec is saying.

Reply:
Here is how it's defined: referencing a global is indeed a `ref`. But atomic functions work on pointers, and from IR point of view the argument is a pointer. So something here needs to be checking for `should_deref`, I suppose